sycl/test-e2e/InvokeSimd/Spec/simd_mask_merge.cpp

```diff
@@ -0,0 +1,108 @@
+// TODO: enable on Windows once driver is ready
+// REQUIRES: gpu && linux
+// UNSUPPORTED: cuda || hip
+//
+// Check that full compilation works:
+// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr %s -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+#include <sycl/detail/boost/mp11.hpp>
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+constexpr int VL = 16;
+
+[[intel::device_indirectly_callable]] simd<float, VL>
+SIMD_CALLEE(simd<float, VL> va, simd_mask<float, VL> mask) SYCL_ESIMD_FUNCTION {
+  esimd::simd<float, VL> ret;
+  esimd::simd_mask<VL> emask;
+  for(int i = 0; i < VL; i++)
+    emask[i] = static_cast<bool>(mask[i]);
+  ret.merge(va, !emask);
+  return ret;
+}
+
+int main() {
+  sycl::queue q;
+  auto dev = q.get_device();
+
+  std::cout << "Running on " << dev.get_info<sycl::info::device::name>()
+            << "\n";
+  constexpr unsigned Size = 1024;
+  constexpr unsigned GroupSize = 4 * VL;
+
+  std::array<float, Size> A;
+  std::array<float, Size> C;
+  std::array<bool, Size> M;
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = i;
+    C[i] = 0;
+    M[i] = i % 2;
+  }
+
+  auto ctxt = q.get_context();
+  sycl::buffer<float> ABuf(A);
+  sycl::buffer<float> CBuf(C);
+  sycl::buffer<bool> MBuf(M);
+
+  sycl::range<1> GlobalRange{Size};
+  // Number of workitems in each workgroup.
+  sycl::range<1> LocalRange{GroupSize};
+
+  sycl::nd_range<1> Range(GlobalRange, LocalRange);
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      sycl::accessor A_acc{ABuf, cgh, sycl::read_only};
+      sycl::accessor C_acc{CBuf, cgh, sycl::write_only};
+      sycl::accessor M_acc{MBuf, cgh, sycl::read_only};
+      cgh.parallel_for(Range, [=](nd_item<1> ndi) {
+        sub_group sg = ndi.get_sub_group();
+        uint32_t wi_id = ndi.get_global_linear_id() + sg.get_local_id();
```

Comment:
nit: sg.get_local_id() should always return 0 for ESIMD:
```suggestion
        uint32_t wi_id = ndi.get_global_linear_id();
```

Reply:
this is invokesimd, so that part of the kernel is sycl only, and i just confirmed it is not returning 0 for this test.

Reply:
I think it would result in would result in something like

```
// WI, SG_SIZE == 2
// | 0 , 2 | 2 , 4 |
```

Reply:
the numerics fail if i remove it, let me investigate more into how the data is being broken up into the nd_range, i don't really understand it

Reply:
> it is not returning 0 for this test

Right, my mistake. I did not notice that it is in a regular SYCL (non ESIMD) kernel

Reply:
yeah there is something weird going on, it may even be a real issue, let me investigate, thanks for pushing me on this one guys

Reply:
ok im really dumb, i thought the default simd constructor zeroed all the elements but it doesn't, that was the root cause of the numeric issue. thanks again guys.

```
  /// Default constructor. Values of the constructed object's elements are
  /// undefined.
  simd_obj_impl() = default;
```

Reply:
Sorry, I kind of noticed that, but just assumed that ESIMD default-inits to zero. Should have asked/clarified instead.